v2/api/appconfiguration/customizations/configuration_store_extensions.go

```diff
@@ -0,0 +1,149 @@
+	obj genruntime.ARMMetaObject,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger) ([]*v1.Secret, error) {
+
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	typedObj, ok := obj.(*storage.ConfigurationStore)
+	if !ok {
+		return nil, errors.Errorf("cannot run on unknown resource type %T, expected *appconfiguration.ConfigurationStore", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = typedObj
+
+	hasSecrets := secretsSpecified(typedObj)
+	if !hasSecrets {
+		log.V(Debug).Info("No secrets retrieval to perform as operatorSpec is empty")
+		return nil, nil
+	}
+
+	id, err := genruntime.GetAndParseResourceID(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]armappconfiguration.APIKey)
+	// Only bother calling ListKeys if there are secrets to retrieve
+	if hasSecrets {
+		subscription := armClient.SubscriptionID()
+		// Using armClient.ClientOptions() here ensures we share the same HTTP connection, so this is not opening a new
+		// connection each time through
+		var confClient *armappconfiguration.ConfigurationStoresClient
+		confClient, err = armappconfiguration.NewConfigurationStoresClient(subscription, armClient.Creds(), armClient.ClientOptions())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create new ConfigurationStoresClient")
+		}
+
+		var pager *runtime.Pager[armappconfiguration.ConfigurationStoresClientListKeysResponse]
+		var resp armappconfiguration.ConfigurationStoresClientListKeysResponse
+		pager = confClient.NewListKeysPager(id.ResourceGroupName, obj.AzureName(), nil)
+		resp, err = pager.NextPage(ctx)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to retreive response")
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed listing keys")
+		}
+
+		keys = secretsByName(resp.Value)
+	}
+
+	secretSlice, err := secretsToWrite(typedObj, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	return secretSlice, nil
+}
+
+func secretsSpecified(obj *storage.ConfigurationStore) bool {
+	if obj.Spec.OperatorSpec == nil || obj.Spec.OperatorSpec.Secrets == nil {
+		return false
+	}
+
+	secrets := obj.Spec.OperatorSpec.Secrets
+
+	if secrets.PrimaryConnectionString != nil ||
+		secrets.PrimaryKeyID != nil ||
+		secrets.PrimaryKey != nil ||
+		secrets.SecondaryConnectionString != nil ||
+		secrets.SecondaryKey != nil ||
+		secrets.SecondaryKeyID != nil ||
+		secrets.PrimaryReadOnlyConnectionString != nil ||
+		secrets.PrimaryReadOnlyKeyID != nil ||
+		secrets.PrimaryReadOnlyKey != nil ||
+		secrets.SecondaryReadOnlyConnectionString != nil ||
+		secrets.SecondaryReadOnlyKey != nil ||
+		secrets.SecondaryReadOnlyKeyID != nil {
+		return true
+	}
+
+	return false
+}
+
+func secretsByName(keys []*armappconfiguration.APIKey) map[string]armappconfiguration.APIKey {
```

Comment:
Given the need to iterate through the pager, noted above, I suggest this method should accept a map that gets populated, rather than creating a new one each time.

Reply:
Done!